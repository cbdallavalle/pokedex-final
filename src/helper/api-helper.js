export const fetchPokemon = async(url) => {
  const rootUrl = `http://localhost:3001/`;
  try {
    const response = await fetch(`${rootUrl}${url}`)
    if(response.status <= 300) {
      const json = await response.json();
      return json
    } else {
      throw new Error ('unable to get pokemon :(')
    }
  } catch(error) {
    throw new Error('unable to get pokemon :(')
  }
}

export const getPokemonTypes = async() => {
  const json = await fetchPokemon(`types`);
  return cleanPokemonTypes(json)
}

export const cleanPokemonTypes = rawTypes => {
  return rawTypes.reduce( (accu, type) => {
    accu[type.name] = type.pokemon
    return accu
  }, {})
}

export const getPokemon = async(pokemonToFetch, type) => {
  const pokemon = pokemonToFetch.map( async id => {
    const json = await fetchPokemon(`pokemon/${id}`)
    return cleanPokemon(json, type)
  })

  return await Promise.all(pokemon)
}

export const cleanPokemon = async(pokemon, type) => ({
    weight: pokemon.weight,
    name: pokemon.name,
    icon: pokemon.sprites.front_default,
    type
})