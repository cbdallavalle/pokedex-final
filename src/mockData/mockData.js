export const mockData = {
  mockPokeTypes: {
    normal: ['16', '17', '18'],
    fighting: ['56', '57', '58'],
    flying: ['21', '22', '23']
  },
  mockRawPokeTypes: [
    {id: "1", name: "normal", pokemon: ['16', '17', '18'] },
    {id: "2", name: "fighting", pokemon: ['56', '57', '58'] },
    {id: "3", name: "flying", pokemon: ['21', '22', '23'] }
  ],
  expectedStoreTypes: {
    type: 'UPLOAD_TYPES',
    pokeTypes: {
    normal: ['16', '17', '18'],
    fighting: ['56', '57', '58'],
    flying: ['21', '22', '23']
    }
  },
  mockPokemonToDisplay: [
    {weight: 18, name: "pidgey", icon: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/16.png", type: "normal"},
    {weight: 300, name: "pidgeotto", icon: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/17.png", type: "normal"}
  ],
  mockRawPokemonData: {
    id: "16",
    name: "pidgey",
    sprites: {
      back_default: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/16.png",
      front_default: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/16.png"
    },
    type: "1",
    weight: 18
  },
  mockCleanPokemonData: [
    {
      weight: 18,
      name: "pidgey",
      icon: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/16.png",
      type: "flying"
    }
  ]
}