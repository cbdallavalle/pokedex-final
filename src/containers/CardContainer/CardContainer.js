import React, { Component } from 'react';
import loadingGif from '../../assets/loading.gif';
import { connect } from 'react-redux';
import { Card } from '../../components/Card/Card';
import * as api from '../../helper/api-helper';

export class CardContainer extends Component {
  constructor(){
    super();
    this.state = {}
  }

  displayLoadingGif = () => {
   return !Object.keys(this.props.pokeTypes).length ? <img src={ loadingGif } alt="loading" /> : <div></div>
  }

  handleClick = async (type) => {
    if(!this.state[type]) {
      const pokemonToFetch = this.props.pokeTypes[type];
      const result = await api.getPokemon(pokemonToFetch, type);
      this.setState({ [type]: result })
    }
  }

  displayCards = () => {
    return Object.keys(this.props.pokeTypes).map( type => {
      return (
        <Card 
          type={ type } 
          handleClick={ this.handleClick }
          pokemonToDisplay={this.state[type]}
        />
      )
    })
  }

  render() {
    console.log(this.state)
    return (
      <section className="CardContainer">
        { this.displayLoadingGif() }
        { this.displayCards() }
      </section>
    )
  }
}

export const mapStateToProps = state => ({
  pokeTypes: state.pokeTypes,
})

export default connect(mapStateToProps, null)(CardContainer);


// = ({ pokeTypes }) => {
//   const loading = !Object.keys(pokeTypes).length ? <img src={loadingGif} alt="loading" /> : <div>LOADED</div>
//   const { bug, fighting, flying, ghost, ground, normal, poison, rock, steel } = pokeTypes;

//   return (
//     <section className="CardContainer">
//       { loading }
//     </section>
//   )
// }