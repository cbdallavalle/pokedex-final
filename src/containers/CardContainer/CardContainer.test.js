import React from 'react';
import { shallow } from 'enzyme';
import { CardContainer, mapStateToProps } from './CardContainer';
import { mockData } from '../../mockData/mockData';
import * as api from '../../helper/api-helper';

describe('CardContainer', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<CardContainer pokeTypes={ mockData.mockPokeTypes } />)
  })

  it('should start with an empty string for error as a default state', () => {
    
    expect(wrapper.state()).toEqual({error: ''})
  })

  //handleClick
  it('handleClick should update state with an error if the fetch returns an error', async() => {
   window.fetch = jest.fn().mockImplementation( () => Promise.resolve({
      status: 500,
      error: new Error('unable to get pokemon :(')
    }))

    await wrapper.instance().handleClick('flying');
    
    expect(wrapper.state().error).toEqual('unable to get pokemon :(')
  })

  it('handleClick should update state if that type of state does not exist', async() => {
    api.getPokemon = () => mockData.mockPokemonToDisplay;
    
    await wrapper.instance().handleClick('flying');
    
    expect(wrapper.state().flying).toEqual(mockData.mockPokemonToDisplay)
  })

  it('handleClick should not update state if that type of state does exist', async() => {
    wrapper.setState({ flying: 'pidgeotto is super cool' })
    api.getPokemon = () => mockData.mockPokemonToDisplay;
    
    await wrapper.instance().handleClick('flying');
    
    expect(wrapper.state().flying).toEqual('pidgeotto is super cool')
  })

  //displayLoadingGif & displayCards
  it('the loading Gif should display if pokeType is an empty object but no cards should display', () => {
    wrapper = shallow(<CardContainer pokeTypes={ {} } />)
    
    expect(wrapper).toMatchSnapshot();
  })

  it('the loading Gif should not display if pokeType exists but the cards should display', () => {
    
    expect(wrapper).toMatchSnapshot();
  })
})

describe('mapStateToProps', () => {
  
  it('should take in state and return an object with pokeTypes as a key', () => {
    const defaultState = {
      pokeTypes: mockData.mockPokeTypes
    }
    const expected = {
      pokeTypes: mockData.mockPokeTypes
    }

    expect(mapStateToProps(defaultState)).toEqual(expected)
  })
})