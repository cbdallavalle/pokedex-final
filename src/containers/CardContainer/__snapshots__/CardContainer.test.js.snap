// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContainer the loading Gif should display if pokeType is an empty object but no cards should display 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CardContainer
    pokeTypes={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <img
          alt="loading"
          src="loading.gif"
        />,
        Array [],
        "",
      ],
      "className": "CardContainer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "alt": "loading",
          "src": "loading.gif",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      "",
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <img
            alt="loading"
            src="loading.gif"
          />,
          Array [],
          "",
        ],
        "className": "CardContainer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "loading",
            "src": "loading.gif",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "",
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`CardContainer the loading Gif should not display if pokeType exists but the cards should display 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CardContainer
    pokeTypes={
      Object {
        "fighting": Array [
          "56",
          "57",
          "58",
        ],
        "flying": Array [
          "21",
          "22",
          "23",
        ],
        "normal": Array [
          "16",
          "17",
          "18",
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        Array [
          <Card
            handleClick={[Function]}
            pokemonToDisplay={undefined}
            type="normal"
          />,
          <Card
            handleClick={[Function]}
            pokemonToDisplay={undefined}
            type="fighting"
          />,
          <Card
            handleClick={[Function]}
            pokemonToDisplay={undefined}
            type="flying"
          />,
        ],
        "",
      ],
      "className": "CardContainer",
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "class",
        "props": Object {
          "handleClick": [Function],
          "pokemonToDisplay": undefined,
          "type": "normal",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "handleClick": [Function],
          "pokemonToDisplay": undefined,
          "type": "fighting",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "handleClick": [Function],
          "pokemonToDisplay": undefined,
          "type": "flying",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "",
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          Array [
            <Card
              handleClick={[Function]}
              pokemonToDisplay={undefined}
              type="normal"
            />,
            <Card
              handleClick={[Function]}
              pokemonToDisplay={undefined}
              type="fighting"
            />,
            <Card
              handleClick={[Function]}
              pokemonToDisplay={undefined}
              type="flying"
            />,
          ],
          "",
        ],
        "className": "CardContainer",
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "handleClick": [Function],
            "pokemonToDisplay": undefined,
            "type": "normal",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "handleClick": [Function],
            "pokemonToDisplay": undefined,
            "type": "fighting",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "handleClick": [Function],
            "pokemonToDisplay": undefined,
            "type": "flying",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "",
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
